src/Network/Wai/Middleware/OpenApi.hs

```diff
@@ -0,0 +1,71 @@
+module Network.Wai.Middleware.OpenApi
+  ( Settings (..)
+  , defaultSettings
+  , ValidationErrors (..)
+  , ValidationError
+  , validateRequestBody
+  , validateResponseBody
+  ) where
+
+import Prelude
+
+import Data.ByteString.Builder (toLazyByteString)
+import Data.ByteString.Lazy qualified as BSL
+import Data.IORef (atomicModifyIORef, newIORef, readIORef)
+import Network.Wai (Middleware, Request, Response)
+import Network.Wai qualified as Wai
+import Network.Wai.Middleware.OpenApi.Schema
+import Network.Wai.Middleware.OpenApi.Settings
+import Network.Wai.Middleware.OpenApi.Validate
+import Network.Wai.Middleware.OpenApi.ValidationError
+
+validateRequestBody :: Settings -> Middleware
+validateRequestBody settings app request respond =
+  validate
+    ValidateOptions
+      { settings
+      , request
+      , schema = lookupRequestSchema settings.spec settings.pathMap request
+      , getBody = getRequestBody request
+      , respondActual = \case
+          Nothing -> app request respond
+          Just newReq -> app newReq respond
+      , respondErrors = respond . clientErrorResponse
+      }
+
+validateResponseBody :: Settings -> Middleware
+validateResponseBody settings app request respond =
+  app request $ \response -> do
+    let status = Wai.responseStatus response
+    validate
+      ValidateOptions
+        { settings
+        , request
+        , schema = lookupResponseSchema status settings.spec settings.pathMap request
+        , getBody = (,()) <$> getResponseBody response
+        , respondActual = \_ -> respond response
+        , respondErrors = respond . serverErrorResponse
+        }
+
+-- | Strictly consume the request body, then mark it as un-consumed
+--
+-- <https://hackage.haskell.org/package/wai-middleware-validation-0.1.0.2/docs/src/Network.Wai.Middleware.Validation.html#getRequestBody>
+getRequestBody :: Request -> IO (BSL.ByteString, Request)
```

Comment:
Having to parse (I know this doesn't parse but we do when we validate) the request twice (and once for the response) for all requests scares me a bit. Should we build in sampling to this library?

Reply:
> Should we build in sampling to this library?

Yes.

Should we do it... before merging this PR? Before we release a version? Before we ourselves use it anywhere? Before we ourselves us it in a busy API? I actually don't really have an opinion; can you tell me?

Reply:
I'm really not sure either- my question was more if that's something this library *should* handle or if you saw that as something that a user of this lib would manage.

Reply:
Well that's a good question. Managing it from the outside would be more work than us managing it internally, **but** it's also an entirely independent concern.

```hs
sampledMiddleware :: Int -> Middleware -> Middleware
sampledMiddleware percent m app request respond = do
  enabled <- (<= percent) <$> randomRs (0, 100)
  if enabled
    then m app request respond
    else app request respond
```

Seems like this should exist somewhere somehow, then we don't have to do it here.

Reply:
I'm going to go ahead and document the performance concern in the README and show this example middleware. We can decide later if we should upstream such a thing to `wai-extra` (where I think it'd belong).